service-base/src/main/java/org/eclipse/hono/service/AbstractAdapterConfig.java

```diff
@@ -134,14 +141,24 @@ protected void setCollaborators(
             adapter.setCommandConsumerFactory(commandConsumerFactory(adapterProperties, samplerFactory, commandRouterClient));
         }
 
+        final KafkaProducerConfigProperties kafkaProducerConfig = kafkaProducerConfig();
+        if (kafkaProducerConfig.getProducerConfig() == null) {
+            // look up via bean factory is not possible because EventSender and TelemetrySender are implemented by
+            // ProtonBasedDownstreamSender
+            adapter.setEventSender(downstreamEventSender(samplerFactory, adapterProperties));
+            adapter.setTelemetrySender(downstreamTelemetrySender(samplerFactory, adapterProperties));
+        } else {
+            final CachingKafkaProducerFactory<String, Buffer> kafkaProducerFactory = kafkaProducerFactory();
+            adapter.setEventSender(downstreamEventKafkaSender(kafkaProducerFactory, kafkaProducerConfig));
+            adapter.setTelemetrySender(downstreamTelemetryKafkaSender(kafkaProducerFactory, kafkaProducerConfig));
+        }
+
```

Comment:
I added #2326. This simplifies this part to the following, by leveraging conditional bean creation:  
```java
// look up client via bean factory in order to take advantage of conditional bean instantiation based
// on config properties
adapter.setEventSender(context.getBean(EventSender.class));
adapter.setTelemetrySender(context.getBean(TelemetrySender.class));
```

Additionally, I can then change the class `KafkaProducerConfig` to no longer return `null`.

@sophokles73 WDYT?